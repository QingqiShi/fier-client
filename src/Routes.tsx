import React, { Suspense, lazy } from 'react';
import { Redirect, Route, Switch } from 'react-router';
import PageLoadIndicator from 'components/base/PageLoadIndicator';
import settings from 'stores/settings';
import user from 'stores/user';
import useRoute from 'hooks/useRoute';

const LazyBottomNav = lazy(() =>
  import(/* webpackChunkName: "user" */ 'components/app/BottomNav')
);
const LazyDashboard = lazy(() =>
  import(/* webpackChunkName: "user" */ 'components/views/Dashboard')
);
const LazyActivity = lazy(() =>
  import(/* webpackChunkName: "user" */ 'components/views/Activity')
);
const LazyCharts = lazy(() =>
  import(/* webpackChunkName: "user" */ 'components/views/Charts')
);
const LazyWallets = lazy(() =>
  import(/* webpackChunkName: "user" */ 'components/views/Wallets')
);

const LazyLogin = lazy(() =>
  import(/* webpackChunkName: "guest" */ 'components/views/Login')
);
const LazyRegister = lazy(() =>
  import(/* webpackChunkName: "guest" */ 'components/views/Register')
);

function Routes() {
  const [{ isLoggedIn }] = user.useStore();
  const [{ locale }] = settings.useStore();

  const { routeLocale, routePath, redirect, getPath } = useRoute();

  if (routeLocale !== locale) {
    redirect(routePath, locale);
    return null;
  }

  return (
    <Suspense fallback={<PageLoadIndicator />}>
      {isLoggedIn ? (
        <>
          <Switch>
            <Route component={LazyDashboard} path={getPath('/dashboard')} />
            <Route component={LazyActivity} path={getPath('/activity')} />
            <Route component={LazyCharts} path={getPath('/charts')} />
            <Route component={LazyWallets} path={getPath('/wallets')} />
            <Route render={() => <Redirect to={getPath('/dashboard')} />} />
          </Switch>
          <LazyBottomNav />
        </>
      ) : (
        <Switch>
          <Route component={LazyLogin} path={getPath('/login')} />
          <Route component={LazyRegister} path={getPath('/register')} />
          <Route render={() => <Redirect to={getPath('/login')} />} />
        </Switch>
      )}
    </Suspense>
  );
}

export default Routes;
