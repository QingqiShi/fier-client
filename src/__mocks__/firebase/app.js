/* ======
  Error Mocks
====== */
let error = false;
export const mockError = (e) => {
  error = e;
};

/* ======
  Auth Mocks
====== */

let authStateChangeCallback = () => {};

const authObj = {
  createUserWithEmailAndPassword: jest.fn(() =>
    error
      ? Promise.reject(error)
      : Promise.resolve({ user: authObj.currentUser })
  ),
  signInWithEmailAndPassword: jest.fn(() =>
    error ? Promise.reject(error) : Promise.resolve()
  ),
  onAuthStateChanged: jest.fn((callback) => {
    authStateChangeCallback = callback;
  }),
  signOut: jest.fn(() => (error ? Promise.reject(error) : Promise.resolve())),
  currentUser: null,
  languageCode: '',
};

export const mockAuthState = (user) => {
  authStateChangeCallback(user);
  authObj.currentUser = user;
};

export const auth = jest.fn(() => authObj);

auth.EmailAuthProvider = {
  credential: jest.fn(),
};

/* ======
  Firestore Mocks
====== */

let states = {};
let snapshotChangeCallbacks = {};

export const mockDocSnapshot = (path, data) => {
  states[path] = data;
  if (path in snapshotChangeCallbacks) {
    snapshotChangeCallbacks[path]({
      exists: !!data,
      data: () => data,
    });
  }
};

export const mockQuerySnapshot = (path, data) => {
  states[path] = data;
  if (path in snapshotChangeCallbacks) {
    snapshotChangeCallbacks[path]({
      forEach: (cb) => {
        data.forEach((x, i) =>
          cb({
            data: () => x,
            id: i,
          })
        );
      },
    });
  }
};

let newDocId = 0;
const mockFirestoreDB = {
  doc: jest.fn((path) => ({
    onSnapshot: (handler) => {
      snapshotChangeCallbacks[path] = handler;
    },
    set: (data, opts = {}) => {
      states[path] = opts.merge ? { ...states[path], ...data } : data;
      snapshotChangeCallbacks[path]({
        exists: true,
        data: () => states[path],
      });
    },
    get: () => ({
      exists: true,
      data: () => states[path],
    }),
    _getPath: () => path,
  })),
  collection: jest.fn((path) => {
    const context = { orderBy: '', order: '', limit: 0 };
    const collectionObj = {
      doc: (name) => {
        if (name) {
          return {
            exists: !!states[`${path}/${name}`],
            data: () => states[`${path}/${name}`],
            _getPath: () => `${path}/${name}`,
          };
        }
        return {
          exists: false,
          data: () => undefined,
          _getPath: () => `${path}/${++newDocId}`,
        };
      },
      orderBy: (field, order) => {
        context.orderBy = field;
        context.order = order;
        return collectionObj;
      },
      limit: (limit) => {
        context.limit = limit;
        return collectionObj;
      },
      onSnapshot: (handler) => {
        snapshotChangeCallbacks[path] = handler;
      },
    };
    return collectionObj;
  }),
  runTransaction: jest.fn((callback) => {
    return callback({
      get: (docRef) =>
        Promise.resolve({
          exists: !!states[docRef._getPath()],
          data: () => states[docRef._getPath()],
        }),
      set: (docRef, data, opts = {}) =>
        mockDocSnapshot(
          docRef._getPath(),
          opts.merge ? { ...states[docRef._getPath()], ...data } : data
        ),
      update: (docRef, data) =>
        mockDocSnapshot(docRef._getPath(), {
          ...states[docRef._getPath()],
          ...data,
        }),
    });
  }),
};

export const firestore = jest.fn(() => mockFirestoreDB);

export function clearFirestoreStates() {
  states = {};
  snapshotChangeCallbacks = {};
  newDocId = 0;
}

export function getMockFirestore(path) {
  return states[path];
}
