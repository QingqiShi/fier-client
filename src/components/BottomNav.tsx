import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import BottomNavigation from '@material-ui/core/BottomNavigation';
import BottomNavigationAction from '@material-ui/core/BottomNavigationAction';
import HomeIcon from '@material-ui/icons/Home';
import TimelineIcon from '@material-ui/icons/Timeline';
import AccountBalanceWalletIcon from '@material-ui/icons/AccountBalanceWallet';
import PieChartIcon from '@material-ui/icons/PieChart';
import Paper from '@material-ui/core/Paper';
import { LocaleRoute, createLocaleUrl } from 'libs/locale-router';
import i18n from 'stores/i18n';
import useTexts from 'hooks/useTexts';

const useStyles = makeStyles({
  root: {
    position: 'fixed',
    bottom: 0,
    paddingBottom: 'env(safe-area-inset-bottom)',
    width: '100%'
  }
});

function BottomNav({ value }: { value: string }) {
  const [t] = useTexts();
  const [{ locale }] = i18n.useStore();
  const classes = useStyles();

  return (
    <LocaleRoute
      render={({ history }) => {
        const handleChange = (e: React.ChangeEvent<{}>, newValue: string) => {
          history.push(createLocaleUrl(locale, newValue));
        };

        return (
          <Paper className={classes.root}>
            <BottomNavigation value={value} onChange={handleChange}>
              <BottomNavigationAction
                label={t['DASHBOARD']}
                icon={<HomeIcon />}
                value="/dashboard"
              />
              <BottomNavigationAction
                label={t['ACTIVITY']}
                icon={<TimelineIcon />}
                value="/activity"
              />
              <BottomNavigationAction
                label={t['CHARTS']}
                icon={<PieChartIcon />}
                value="/charts"
              />
              <BottomNavigationAction
                label={t['WALLETS']}
                icon={<AccountBalanceWalletIcon />}
                value="/wallets"
              />
            </BottomNavigation>
          </Paper>
        );
      }}
    />
  );
}

export default BottomNav;
