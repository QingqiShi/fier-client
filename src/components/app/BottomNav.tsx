import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import {
  BottomNavigation,
  BottomNavigationAction,
  Paper
} from '@material-ui/core';
import {
  AccountBalanceWallet,
  Home,
  PieChart,
  Timeline
} from '@material-ui/icons';
import useTexts from 'hooks/useTexts';
import useRoute from 'hooks/useRoute';

const useStyles = makeStyles({
  root: {
    position: 'fixed',
    bottom: 0,
    paddingBottom: 'env(safe-area-inset-bottom)',
    width: '100%'
  }
});

const navigationConfig = [
  { label: 'DASHBOARD', icon: Home, value: '/dashboard' },
  { label: 'ACTIVITY', icon: Timeline, value: '/activity' },
  { label: 'CHARTS', icon: PieChart, value: '/charts' },
  { label: 'WALLETS', icon: AccountBalanceWallet, value: '/wallets' }
];

function BottomNav() {
  const [t] = useTexts();
  const classes = useStyles();
  const { routePath, redirect } = useRoute();

  const selectedNav =
    navigationConfig.find(config =>
      routePath.match(new RegExp(config.value))
    ) || navigationConfig[0];

  const handleChange = (e: React.ChangeEvent<{}>, newValue: string) => {
    redirect(newValue);
  };

  return (
    <Paper className={classes.root}>
      <BottomNavigation value={selectedNav.value} onChange={handleChange}>
        {navigationConfig.map(nav => (
          <BottomNavigationAction
            key={`bottom-nav-${nav.value}`}
            icon={<nav.icon />}
            label={t[nav.label]}
            value={nav.value}
          />
        ))}
      </BottomNavigation>
    </Paper>
  );
}

export default BottomNav;
