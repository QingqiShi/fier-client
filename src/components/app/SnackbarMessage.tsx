import React, { useCallback } from 'react';
import { Button, Slide, Snackbar } from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { TransitionProps } from '@material-ui/core/transitions/transition';
import snackbar from 'stores/snackbar';

function SlideTransition(props: TransitionProps) {
  return <Slide {...props} direction="down" />;
}

function SnackbarMessage() {
  const [
    { type, message, actionLabel, action, isShowing, hideAfter },
    { clearMessage }
  ] = snackbar.useStore();

  const handleSnackbarClose = useCallback(
    (e: React.SyntheticEvent<any, Event>, reason?: string) => {
      if (reason === 'clickaway') return;
      clearMessage();
    },
    [clearMessage]
  );

  return (
    <Snackbar
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'center'
      }}
      autoHideDuration={hideAfter || undefined}
      open={isShowing}
      TransitionComponent={SlideTransition}
      onClose={handleSnackbarClose}
    >
      <Alert
        action={
          action &&
          actionLabel && <Button onClick={action}>{actionLabel}</Button>
        }
        elevation={6}
        severity={type}
        variant="filled"
        onClose={clearMessage}
      >
        {message}
      </Alert>
    </Snackbar>
  );
}

export default SnackbarMessage;
