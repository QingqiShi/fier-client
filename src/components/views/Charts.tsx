import React from 'react';
import TopNav from 'components/app/TopNav';
import useTexts from 'hooks/useTexts';

function Charts() {
  const [t] = useTexts();
  return (
    <div style={{ padding: '5em 0' }}>
      <TopNav title={t['CHARTS']} />
    </div>
  );
}

export default Charts;
