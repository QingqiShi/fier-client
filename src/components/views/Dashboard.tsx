import React from 'react';
import {
  Card,
  CardActionArea,
  CardContent,
  Fab,
  GridList,
  GridListTile,
  List,
  ListItem,
  ListItemIcon,
  ListItemSecondaryAction,
  ListItemText,
  ListSubheader,
  Theme,
  Typography,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import { Add } from '@material-ui/icons';
import Currency from 'components/base/Currency';
import TopNav from 'components/app/TopNav';
import useTexts from 'hooks/useTexts';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      padding: theme.spacing(13, 2),
    },
    sectionTitle: {
      margin: theme.spacing(5, 0, 1),
    },
    gridListContainer: {
      margin: theme.spacing(0, -2),
      padding: theme.spacing(1, 2),
      overflowX: 'auto',
    },
    gridList: {
      flexWrap: 'nowrap',
      transform: 'translateZ(0)',
      overflow: 'visible',
    },
    gridListTile: {
      overflow: 'visible',
    },
    gridAddButton: {
      height: 'calc(100% - 16px)',
      display: 'flex',
    },
    accountBalance: {
      marginTop: theme.spacing(2),
    },
    emoji: {
      fontSize: theme.typography.h5.fontSize,
      color: theme.palette.text.primary,
    },
    listHeader: { top: 72, backgroundColor: theme.palette.background.default },
    fab: {
      position: 'fixed',
      right: theme.spacing(2),
      bottom: `calc(env(safe-area-inset-bottom) + ${theme.spacing(2)}px)`,
      zIndex: theme.zIndex.speedDial,
    },
  })
);

function Dashboard() {
  const [t] = useTexts();
  const classes = useStyles();

  return (
    <div className={classes.container}>
      <TopNav title={t.APP_NAME} />

      <Typography align="right" color="textSecondary" variant="subtitle1">
        {t['NET_WORTH']}
      </Typography>
      <Currency
        align="right"
        currency="GBP"
        value={1234.56}
        variant={['h3', 'h4']}
      />

      <Typography className={classes.sectionTitle} variant="h6">
        {t['ACCOUNTS']}
      </Typography>
      <div className={classes.gridListContainer}>
        <GridList
          cellHeight={140}
          className={classes.gridList}
          cols={1.5}
          spacing={16}
        >
          {[1].map((i) => (
            <GridListTile
              key={`account-${i}`}
              classes={{
                tile: classes.gridListTile,
              }}
            >
              <Card className={classes.gridAddButton} elevation={4}>
                <CardActionArea>
                  <CardContent>
                    <Typography color="textSecondary" variant="subtitle1">
                      Xxxx xxxxxxx
                    </Typography>
                    <Typography>Xxxxx</Typography>

                    <Currency
                      align="right"
                      currency="GBP"
                      value={1234.56}
                      variant={['h6', 'subtitle1']}
                    />
                  </CardContent>
                </CardActionArea>
              </Card>
            </GridListTile>
          ))}
          <GridListTile
            classes={{
              tile: classes.gridListTile,
            }}
            style={{ paddingRight: 16 }}
          >
            <Card className={classes.gridAddButton} elevation={4}>
              <CardActionArea>
                <CardContent>
                  <Typography align="center" color="textSecondary" variant="h3">
                    +
                  </Typography>
                  <Typography align="center" color="textSecondary">
                    {t.CREATE_ACCOUNT}
                  </Typography>
                </CardContent>
              </CardActionArea>
            </Card>
          </GridListTile>
        </GridList>
      </div>

      <Typography className={classes.sectionTitle} variant="h6">
        {t['TRANSACTIONS']}
      </Typography>
      <List
        component="div"
        subheader={
          <ListSubheader
            className={classes.listHeader}
            component="div"
            disableGutters
          >
            xx/xx/xxxx
          </ListSubheader>
        }
      >
        {['🛍', '🛒', '🚇'].map((emoji, i) => (
          <ListItem key={`first-list-item-${i}`} ContainerComponent="div">
            <ListItemIcon className={classes.emoji}>
              <span>{emoji}</span>
            </ListItemIcon>
            <ListItemText primary="Xxxxx xxx xxxx" secondary="Xxxx x xxxxx" />
            <ListItemSecondaryAction>
              <Currency
                currency="GBP"
                value={-12.34}
                variant={['body1', 'body2']}
              />
            </ListItemSecondaryAction>
          </ListItem>
        ))}
      </List>
      <List
        component="div"
        subheader={
          <ListSubheader
            className={classes.listHeader}
            component="div"
            disableGutters
          >
            xx/xx/xxxx
          </ListSubheader>
        }
      >
        {['🛍', '🛒', '🧾', '🚇', '💸'].map((emoji, i) => (
          <ListItem key={`second-list-item-${i}`} ContainerComponent="div">
            <ListItemIcon className={classes.emoji}>
              <span>{emoji}</span>
            </ListItemIcon>
            <ListItemText primary="Xxxxx xxx xxxx" secondary="Xxxx x xxxxx" />
            <ListItemSecondaryAction>
              <Currency
                currency="GBP"
                value={-12.34}
                variant={['body1', 'body2']}
              />
            </ListItemSecondaryAction>
          </ListItem>
        ))}
      </List>

      <Fab className={classes.fab} color="primary">
        <Add />
      </Fab>
    </div>
  );
}

export default Dashboard;
