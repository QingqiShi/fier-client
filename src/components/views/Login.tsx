import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import { Typography } from '@material-ui/core';
import OverlayContainer from 'components/base/OverlayContainer';
import SignInFields from 'components/app/SignInFields';
import LanguageSelector from 'components/app/LanguageSelector';
import useTexts from 'hooks/useTexts';
import useFormInput from 'hooks/useFormInput';
import useFirebaseAuth from 'hooks/useFirebaseAuth';

function Login() {
  const [t] = useTexts();

  const { user, signIn } = useFirebaseAuth();
  const [loading, setLoading] = useState(false);

  const emailControl = useFormInput('');
  const passwordControl = useFormInput('');

  function handleSubmit(e: React.FormEvent) {
    e.preventDefault();

    setLoading(true);
    const email = emailControl[0];
    const password = passwordControl[0];
    signIn({ email, password }).then(user => {
      if (!user) {
        setLoading(false);
      }
    });

    return false;
  }

  if (user.isLoggedIn) {
    return <Redirect to="/dashboard" />;
  }

  return (
    <OverlayContainer maxWidth="xs">
      <Typography component="h1" variant="h3" gutterBottom>
        {t['LOGIN']}
      </Typography>
      <form autoComplete="on" onSubmit={handleSubmit}>
        <SignInFields
          emailControl={emailControl}
          loading={loading}
          passwordControl={passwordControl}
        />
      </form>
      <LanguageSelector position="fixed" />
    </OverlayContainer>
  );
}

export default Login;
