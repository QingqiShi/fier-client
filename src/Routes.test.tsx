import React from 'react';
import { act, fireEvent, waitFor } from '@testing-library/react';
import createMockRaf, { MockRaf } from '@react-spring/mock-raf';
import { FrameLoop, Globals } from '@react-spring/web';
import {
  clearAuthListeners,
  mockAuthUser,
  mockFirestore,
  render
} from 'testUtils';
import SnackbarMessage from 'components/app/SnackbarMessage';
import snackbar from 'stores/snackbar';
import Routes from './Routes';

// Load files upfront to suppress lazy load
import('components/app/BottomNav');
import('components/views/Dashboard');
import('components/views/Activity');
import('components/views/Charts');
import('components/views/Wallets');
import('components/views/Login');
import('components/views/Register');

let mockRaf: MockRaf;
beforeEach(() => {
  mockRaf = createMockRaf();
  Globals.assign({
    now: mockRaf.now,
    requestAnimationFrame: mockRaf.raf,
    cancelAnimationFrame: mockRaf.cancel,
    frameLoop: new FrameLoop()
  });
});

beforeEach(() => {
  clearAuthListeners();
  jest.clearAllMocks();
});

test('renders guest user', async () => {
  const { findAllByText } = render(<Routes />, { userAndSettings: true });
  act(() => void mockAuthUser(null));
  expect((await findAllByText('Log in')).length).toBeGreaterThan(0);
});

test('redirects to user locale', async () => {
  const { findAllByText } = render(<Routes />, { userAndSettings: true });

  act(() => void mockAuthUser({ uid: 'testid' }));
  act(() =>
    mockFirestore('settings/testid', { locale: 'zh', categories: [1] })
  );

  expect((await findAllByText('概览')).length).toBeGreaterThan(0);
});

test('renders logged in user', async () => {
  const { getByTestId } = render(<Routes />, { userAndSettings: true });

  act(() => void mockAuthUser({ uid: 'testid' }));
  act(() =>
    mockFirestore('settings/testid', { locale: 'en', categories: [1] })
  );

  await waitFor(() => {
    getByTestId('topnav-profile');
  });
});

test('show setup modal if categories has length 0', async () => {
  const { findByText } = render(<Routes />, { userAndSettings: true });

  act(() => void mockAuthUser({ uid: 'testid' }));
  act(() => mockFirestore('settings/testid', { locale: 'en', categories: [] }));
  act(() => mockRaf.flush());

  expect(await findByText('Categories')).toBeInTheDocument();
});

test('show snackbar message when updates available', async () => {
  const { findByText, getByText } = render(
    <div>
      <Routes />
      <SnackbarMessage />
    </div>,
    { userAndSettings: true, stores: [snackbar] }
  );

  act(() => void mockAuthUser({ uid: 'testid' }));
  act(() => mockFirestore('settings/testid', { locale: 'en', categories: [] }));
  act(() => mockRaf.flush());

  // Mock service worker updated callback
  act(() => window.swStates.callback && window.swStates.callback());
  expect(await findByText(/New update available/)).toBeInTheDocument();

  fireEvent.click(getByText('Update Now'));
  expect(window.swStates.updateAndReload).toHaveBeenCalled();
});

test('show snackbar message when update event was missed', async () => {
  // Mock service worker update have happened already
  window.swStates.updated = true;

  const { findByText, getByText } = render(
    <div>
      <Routes />
      <SnackbarMessage />
    </div>,
    { userAndSettings: true, stores: [snackbar] }
  );

  act(() => void mockAuthUser({ uid: 'testid' }));
  act(() => mockFirestore('settings/testid', { locale: 'en', categories: [] }));
  act(() => mockRaf.flush());

  expect(await findByText(/New update available/)).toBeInTheDocument();

  fireEvent.click(getByText('Update Now'));
  expect(window.swStates.updateAndReload).toHaveBeenCalled();
});
