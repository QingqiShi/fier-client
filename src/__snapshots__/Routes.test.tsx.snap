// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders guest user 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 makeStyles-overlay-23"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXs"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom"
      >
        Log in
      </h1>
      <form
        autocomplete="on"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
            data-shrink="false"
          >
            Email
            <span
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
              â€‰*
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              autocomplete="email"
              class="MuiInputBase-input MuiInput-input"
              name="email"
              required=""
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
            data-shrink="false"
          >
            Password
            <span
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
              â€‰*
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              autocomplete="new-password"
              class="MuiInputBase-input MuiInput-input"
              required=""
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-loginButton-88 MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Log in
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-link-180 MuiTypography-body2 MuiTypography-colorPrimary"
        >
          Don't have an account? Sign up
        </a>
      </form>
      <div
        class="makeStyles-languageContainer-188 makeStyles-languageContainer-189"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby=" mui-component-select-language"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
            id="mui-component-select-language"
            role="button"
            tabindex="0"
          >
            <span
              class="makeStyles-emoji-187"
              role="img"
            >
              ðŸ‡ºðŸ‡¸
            </span>
          </div>
          <input
            name="language"
            type="hidden"
            value="en"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders logged in user 1`] = `
<DocumentFragment>
  <div
    style="padding: 5em 0px;"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation0 MuiAppBar-root MuiAppBar-positionFixed makeStyles-appBar-246 makeStyles-appBar-250 MuiAppBar-colorPrimary mui-fixed"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular makeStyles-toolBar-247 MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart"
          data-testid="topnav-profile"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-245 MuiAvatar-colorDefault"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h5
          class="MuiTypography-root makeStyles-typography-248 MuiTypography-h5 MuiTypography-colorTextPrimary"
        >
          Dashboard
        </h5>
        <div
          class="makeStyles-spacer-249"
        />
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-edgeEnd"
          data-testid="topnav-add"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-348 MuiPaper-rounded"
  >
    <div
      class="MuiBottomNavigation-root"
    >
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          <span
            class="MuiBottomNavigationAction-label Mui-selected"
          >
            Dashboard
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <defs>
              <path
                d="M0 0h24v24H0V0z"
                id="a"
              />
            </defs>
            <path
              d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
            />
          </svg>
          <span
            class="MuiBottomNavigationAction-label MuiBottomNavigationAction-iconOnly"
          >
            Activity
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 2v20c-5.07-.5-9-4.79-9-10s3.93-9.5 9-10zm2.03 0v8.99H22c-.47-4.74-4.24-8.52-8.97-8.99zm0 11.01V22c4.74-.47 8.5-4.25 8.97-8.99h-8.97z"
            />
          </svg>
          <span
            class="MuiBottomNavigationAction-label MuiBottomNavigationAction-iconOnly"
          >
            Charts
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root MuiBottomNavigationAction-iconOnly"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 18v1c0 1.1-.9 2-2 2H5c-1.11 0-2-.9-2-2V5c0-1.1.89-2 2-2h14c1.1 0 2 .9 2 2v1h-9c-1.11 0-2 .9-2 2v8c0 1.1.89 2 2 2h9zm-9-2h10V8H12v8zm4-2.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
            />
          </svg>
          <span
            class="MuiBottomNavigationAction-label MuiBottomNavigationAction-iconOnly"
          >
            Wallets
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
